File: samples/bot-daily-task-reminder/nodejs/models/taskModuleResponseFactory.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

/**
 * TaskModuleResponseFactory is a utility class for creating task module responses.
 * It provides methods for generating task responses, either as a message or a continuation.
 */
class TaskModuleResponseFactory {
    /**
     * Creates a task module response based on the input type.
     * If the input is a string, a message type response is created.
     * If the input is an object, a continue type response is created.
     * 
     * @param {string|Object} taskModuleInfoOrString - The task module info or string value to generate the response.
     * @returns {Object} The generated task module response.
     */
    static createResponse(taskModuleInfoOrString) {
        if (typeof taskModuleInfoOrString === 'string') {
            return {
                task: {
                    type: 'message',
                    value: taskModuleInfoOrString
                }
            };
        }

        return {
            task: {
                type: 'continue',
                value: taskModuleInfoOrString
            }
        };
    }

    /**
     * A wrapper for createResponse. It returns a task module response based on the provided taskInfo.
     * 
     * @param {Object} taskInfo - The task info to generate the response.
     * @returns {Object} The task module response.
     */
    static toTaskModuleResponse(taskInfo) {
        return TaskModuleResponseFactory.createResponse(taskInfo);
    }
}

module.exports.TaskModuleResponseFactory = TaskModuleResponseFactory;

File: samples/bot-task-module/python/src/models/task_module_response_factory.py
# Copyright (c) Microsoft Corp. All rights reserved.
# Licensed under the MIT License.

from typing import Union, Any


class TaskModuleTaskInfo:
    """
    Task Module task info for configuring the task module dialog.
    """

    def __init__(self):
        self.title: str = None
        self.height: int = None
        self.width: int = None
        self.url: str = None
        self.fallback_url: str = None
        self.card: Any = None

    def to_dict(self) -> dict:
        """Convert to dictionary for API response."""
        result = {}
        if self.title:
            result["title"] = self.title
        if self.height:
            result["height"] = self.height
        if self.width:
            result["width"] = self.width
        if self.url:
            result["url"] = self.url
        if self.fallback_url:
            result["fallbackUrl"] = self.fallback_url
        if self.card:
            result["card"] = self.card
        return result


class TaskModuleResponseFactory:
    """
    A factory class to create TaskModuleResponse objects.
    """

    @staticmethod
    def create_response(value: Union[str, TaskModuleTaskInfo]) -> dict:
        """
        Creates a TaskModuleResponse based on the provided value.

        :param value: A string or TaskModuleTaskInfo object.
        :return: A TaskModuleResponse dict.
        """
        if isinstance(value, TaskModuleTaskInfo):
            return {
                "task": {
                    "type": "continue",
                    "value": value.to_dict()
                }
            }
        return {
            "task": {
                "type": "message",
                "value": value
            }
        }

    @staticmethod
    def to_task_module_response(task_info: TaskModuleTaskInfo) -> dict:
        """
        Converts TaskModuleTaskInfo to a TaskModuleResponse.

        :param task_info: The TaskModuleTaskInfo object.
        :return: A TaskModuleResponse dict.
        """
        return TaskModuleResponseFactory.create_response(task_info)
